board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c

```diff
@@ -87,6 +106,264 @@ int board_early_init_f(void)
+		printf("%x", fit_hash->value[i]);
+	printf("\n");
+
+	if (m4_upgrade_required(fit_hash) == false) {
+		printf("M4: upgrade not required\n");
+		return 0;
+	}
+
+	return 1;
+}
+
+static int m4_get_fit_node(const void *fit)
+{
+	int images_noffset;
+	const char *name;
+	int noffset;
+	int ndepth;
+	int count;
+
+	images_noffset = fdt_path_offset(fit, FIT_IMAGES_PATH);
+	if (images_noffset < 0) {
+		printf("M4: Can't find images parent node '%s' (%s)\n",
+		       FIT_IMAGES_PATH, fdt_strerror(images_noffset));
+		return -EINVAL;
+	}
+
+	for (ndepth = 0, count = 0,
+	     noffset = fdt_next_node(fit, images_noffset, &ndepth);
+			(noffset >= 0) && (ndepth > 0);
+			noffset = fdt_next_node(fit, noffset, &ndepth)) {
+
+		if (ndepth == 1) {
+			name = fit_get_name(fit, noffset, NULL);
+			if (!strncmp(name, "m4", 2))
+				return noffset;
+ 			count++;
+		}
+	}
+
+	printf("M4: Can't find the FPGA M4 node\n");
+
+	return -EINVAL;
+}
+
+static void m4_boot(struct spi_flash *flash)
+{
+	int hash_len = 0, tee_hash_len = 0, tee_payload_len = 0;
+	struct mtd_info *mtd = 	&flash->mtd;
+	uint8_t tee_hash[FIT_MAX_HASH_LEN];
+	uint8_t hash[FIT_MAX_HASH_LEN];
+	size_t len;
+
+	if (optee_get_tee_hash(tee_hash, &tee_hash_len, &tee_payload_len)) {
+		printf("M4: cant boot, TEE not accessible\n");
+		return;
+	}
+
+	/* clear all memory to prevent against possible attacks */
+	memset((void *) M4_BASE, 0x00, M4_SIZE);
+
+	/* read the image */
+	len = round_up(tee_payload_len, mtd->writesize);
+	if (spi_flash_read(flash, 0, len, (void *) M4_BASE)) {
+		printf("M4: Failed to read from flash, can't boot\n");
+		return;
+	}
+
+	printf("\tM4 Entry: 0x%x\n", M4_ENTRY(M4_BASE));
+	printf("\tM4 Sram : 0x%x\n", M4_BASE);
+	printf("\tM4 Size : 0x%x\n", tee_payload_len);
+
+	/* do not boot if the FPGA is corrupted (ie, sha different than TEE's) */
+	if (calculate_hash((const void *) M4_BASE, tee_payload_len, "sha256",
+			   hash, &hash_len)) {
+		printf("M4: cant boot, unsupported hash algorithm\n");
+		return;
+	}
+
+	if (hash_len != tee_hash_len) {
+		printf("M4: cant boot, invalid hash length\n");
+		return;
+	} else if (memcmp(hash, tee_hash, tee_hash_len)) {
+		printf("M4: cant boot, invalid hash values\n");
+		return;
+	} else
+		printf("   M4 Firmware SHA validated, booting\n");
+
+	writel(M4_ENTRY(M4_BASE), SIM0_RBASE + 0x70);
+}
+
+/* As part of the boot sequence, SPL checks all images in the FIT; we chose
+ * to register the M4 firmware as an FPGA image to benefit from this loading
+ * process: ie, before booting the kernel, fpga_loadbistream will get executed
+ * allowing the M4 firmware to be loaded and run.
+ */
+int fpga_loadbitstream(int devnum, char *fpgadata, size_t size,
+		       bitstream_type bstype)
+{
+	struct spi_flash *flash;
+	struct hash fit_hash;
+	int noffset, ret;
+
+	if (get_boot_mode() == DUAL_BOOT) {
+		/* in DUAL mode, the M4 should be already running */
+		printf("M4: Cant upgrade or boot M4 in dual boot mode\n");
+		return 0;
```

Comment:
Wonder if we should return an error here, as otherwise it might give the wrong impression that it was able to parse and load the m4 content from fit, which is not the case.

Reply:
we cant return an error here since at this point we don't know yet if the user wanted to upgrade. but we can add this additional check to where the upgrade is supposed to initiate and then raise the error there?

Reply:
This function will get called when parsing the fit image (and if the fit includes the fpga entry), so we should cover all the cases you suggested here.

First is what to do if boot mode is not single boot? I would suggest aborting as the user is doing something wrong anyway (loading a fit that includes fpga but booting in a scenario in which u-boot is unable to update m4).

Then we need to make sure the M4 is in a proper state, to cover the A7 reboot scenario. We can always have linux to message M4 during a normal shutdown process, but we will have to cover the unexpected reboot one as you mentioned, as at that point M4 is going to be running while we try to boot/update it. The same logic will also apply to rollback indeed, as we can have failures at both M4 and A7.